test/framework/src/test/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertionsTests.java

```diff
@@ -52,4 +59,161 @@ public void writeTo(StreamOutput out) throws IOException {
+                original.value(randomAlphaOfLength(10));
+            }
+            original.endArray();
+        }
+        original.endObject();
+
+        XContentParser parser = XContentHelper.createParser(NamedXContentRegistry.EMPTY, original.bytes(), original.contentType());
+        XContentBuilder copy = JsonXContent.contentBuilder();
+        parser.nextToken();
+        XContentHelper.copyCurrentStructure(copy.generator(), parser);
+        assertToXContentEquivalent(original.bytes(), copy.bytes(), original.contentType());
+    }
+
+    public void testAssertXContentEquivalentErrors() throws IOException {
+        {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            {
+                builder.startObject("foo");
+                {
+                    builder.field("field1", "value1");
+                    builder.field("field2", "value2");
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+
+            XContentBuilder otherBuilder = JsonXContent.contentBuilder();
+            otherBuilder.startObject();
+            {
+                otherBuilder.startObject("foo");
+                {
+                    otherBuilder.field("field1", "value1");
+                }
+                otherBuilder.endObject();
+            }
+            otherBuilder.endObject();
+            AssertionError error = expectThrows(AssertionError.class,
+                    () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
+            assertEquals("different field names at [foo]: [field1, field2] vs. [field1] expected:<2> but was:<1>",
+                    error.getMessage());
+        }
+        {
+            XContentBuilder builder = JsonXContent.contentBuilder();
+            builder.startObject();
+            {
+                builder.startObject("foo");
+                {
+                    builder.field("field1", "value1");
+                    builder.field("field2", "value2");
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+
+            XContentBuilder otherBuilder = JsonXContent.contentBuilder();
+            otherBuilder.startObject();
+            {
+                otherBuilder.startObject("foo");
+                {
+                    otherBuilder.field("field1", "value1");
+                    otherBuilder.field("field2", "differentValue2");
+                }
+                otherBuilder.endObject();
+            }
+            otherBuilder.endObject();
+            AssertionError error = expectThrows(AssertionError.class,
+                    () -> assertToXContentEquivalent(builder.bytes(), otherBuilder.bytes(), builder.contentType()));
+            assertEquals("different xContent at [foo/field2] expected:<[v]alue2> but was:<[differentV]alue2>", error.getMessage());
```

Comment:
That part should show which part of a larger xContent object tree the error occurs at. Maybe using dots here like we do in the xContent path syntax in the yaml tests would help? What about `different xContent values at [foo.field2]` and then the expected and actual values?